onlinejudge/service/toph.py

```diff
@@ -0,0 +1,169 @@
+# Python Version: 3.x
+import posixpath
+import string
+import urllib.parse
+from typing import *
+
+import bs4
+import requests
+import re
+
+import onlinejudge.dispatch
+import onlinejudge.implementation.logging as log
+import onlinejudge.implementation.utils as utils
+import onlinejudge.type
+from onlinejudge.type import SubmissionError
+
+
+@utils.singleton
+class TophService(onlinejudge.type.Service):
+    def login(self, get_credentials: onlinejudge.type.CredentialsProvider, session: Optional[requests.Session] = None) -> bool:
+        session = session or utils.new_default_session()
+        url = 'https://toph.co/login'
+        # get
+        resp = utils.request('GET', url, session=session)
+        if resp.url != url:  # redirected
+            log.info('You are already logged in.')
+            return True
+        # parse
+        soup = bs4.BeautifulSoup(resp.content.decode(resp.encoding), utils.html_parser)
+        form = soup.find('form', class_='login-form')
+        log.debug('form: %s', str(form))
+        username, password = get_credentials()
+        form['action'] = '/login' # to avoid KeyError inside form.request method as Toph does not have any defined action
+        form = utils.FormSender(form, url=resp.url)
+        form.set('handle', username)
+        form.set('password', password)
+        # post
+        resp = form.request(session)
+        resp.raise_for_status()
+
+        newResp = utils.request('GET', url, session=session)    # Toph's Location header is not getting the expected value
+        if newResp.url != url:
+            log.success('Welcome, %s.', username)
+            return True
+        else:
+            log.failure('Invalid handle/email or password.')
+            return False
+
+    def is_logged_in(self, session: Optional[requests.Session] = None) -> bool:
+        session = session or utils.new_default_session()
+        url = 'https://toph.co/login'
+        resp = utils.request('GET', url, session=session, allow_redirects=False)
+        return resp.status_code == 302
+
+    def get_url(self) -> str:
+        return 'https://toph.co/'
+
+    def get_name(self) -> str:
+        return 'toph'
+
+    @classmethod
+    def from_url(cls, s: str) -> Optional['TophService']:
+        # example: https://toph.co/
+        # example: http://toph.co/
+        result = urllib.parse.urlparse(s)
+        if result.scheme in ('', 'http', 'https') \
+                and result.netloc == 'toph.co':
+            return cls()
+        return None
+
+class TophProblem(onlinejudge.type.Problem):
+    def __init__(self, slug: str, kind: Optional[str] = None):
```

Comment:
```suggestion
    def __init__(self, problem_id: str, contest_id: Optional[str] = None):
        if contest_id is not None:
            raise NotImplementedError
```

This argument `kind` is for the problems in the contest mode, right?
When we support the contest mode, it seems that this class need to have something like `contest_id` since the urls of problems in contets are like https://toph.co/c/tough-dash-feb-2019/arena#!/p/set-union .
So, we should use such names from now, and mark it as not-implemented explicitly.